jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java

```diff
@@ -122,25 +145,68 @@ public Content.Chunk read()
                         transformedChunk = Content.Chunk.asChunk(transformedChunk.getByteBuffer(), false, transformedChunk);
                 }
 
-                boolean terminated = rawLast && transformedLast;
-                boolean terminalFailure = transformedFailure && transformedLast;
-
-                Content.Chunk result = transformedChunk;
-                transformedChunk = Content.Chunk.next(result);
+                if (transformedFailure && transformedLast)
+                    return afterRead(State.Type.FAILED, transformedChunk);
 
-                if (terminated || terminalFailure)
-                {
-                    finished = true;
-                    release();
-                }
+                if (rawLast && transformedLast)
+                    return afterRead(State.Type.EOF, transformedChunk);
 
-                return result;
+                return afterRead(State.Type.IDLE, transformedChunk);
             }
 
             needsRawRead = rawChunk == null;
         }
     }
 
+    private Content.Chunk afterRead(State.Type targetType, Content.Chunk chunk)
+    {
+        while (true)
+        {
+            State current = state.get();
+            switch (current.type)
+            {
+                case IDLE, EOF, FAILED -> throw new IllegalStateException();
+                case READING ->
+                {
+                    switch (targetType)
+                    {
+                        case IDLE ->
+                        {
+                            if (state.compareAndSet(current, State.IDLE))
+                                return chunk;
+                        }
+                        case FAILED ->
+                        {
+                            if (state.compareAndSet(current, new State.Failed(chunk)))
+                            {
+                                dispose(chunk.getFailure());
+                                return chunk;
+                            }
+                        }
+                        case EOF ->
+                        {
+                            if (state.compareAndSet(current, State.EOF))
+                            {
+                                release();
+                                return chunk;
+                            }
+                        }
+                        default -> throw new IllegalStateException();
+                    }
+                }
+                case FAILING ->
+                {
+                    Content.Chunk failedChunk = ((State.Failing)current).chunk;
+                    if (state.compareAndSet(current, new State.Failed(failedChunk)))
+                    {
+                        dispose(failedChunk.getFailure());
+                        return chunk;
+                    }
```

Comment:
I also don't think this is correct.   Once we have a persistent failure, then any future failures should just be added as suppressed exceptions to the original.

Reply:
I don't follow? This code handles a concurrent call to `fail()` during a call to `read()`. There is only 1 failure, so it cannot be added as suppressed, as it is the first failure.